docs/csharp/csharp-7.md

```diff
@@ -16,64 +16,539 @@ ms.assetid: fd41596d-d0c2-4816-b94d-c4d00a5d0243
+defines the type returned. There is no need for creating a new symbol.
+
+Creating a tuple is more efficient and more productive.
+It is a simpler, lightweight syntax to define a data structure that carries
+more than one value. The example method below returns the minimimum and maximum
+values found in a sequence of integers:
+
+[!code-csharp[TupleReturningMethod](../../samples/snippets/csharp/new-in-7/new-in-7/program.cs#08_TupleReturningMethod "Tuple returning method")]
+
+The declaration for the method provides the names for the fields of the
+tuple that is returned. When you call the method, the return value is a 
+tuple whose fields are `max` and `min`:
+
+[!code-csharp[CallingTupleMethod](../../samples/snippets/csharp/new-in-7/new-in-7/program.cs#09_CallingTupleMethod "Calling a tuple returning method")]
+
+There may be times when you want to unpackage the members of a tuple that
+was returned from a method.  You can do that by declaring separate variables
+for each of the values in the tuple. This is called *deconstructing* the tuple:
+
+[!code-csharp[CallingWithDeconstructor](../../samples/snippets/csharp/new-in-7/new-in-7/program.cs#10_CallingWithDeconstructor "Deconstructing a tuple")]
+
+<!-- Add wildcards here, if they are in C# 7
+-->
 
-### Deconstruction
+You can also provide a simailar deconstruction for any type in .NET. This is
+done by writing a `Deconstruct` method as a member of the class. That
+`Deconstruct` method provides a set of `out` arguments for each of the
+properties you want to extract. Consider
+this `Point` class that provides a deconstructor method that extracts
+the `X` and `Y` coordinates:
+
+[!code-csharp[PointWithDeconstruction](../../samples/snippets/csharp/new-in-7/new-in-7/point.cs#11_PointWithDeconstruction "Point with deconstruction method")]
+ 
+You can extract the individual fields by assigning a tuple to a `Point`:
+
+[!code-csharp[DeconstructPoint](../../samples/snippets/csharp/new-in-7/new-in-7/program.cs#12_DeconstructPoint "Deconstruct a point")]
+
+You are not bound by the names defined in the `Deconstruct` method. You
+can rename the extract variables as part of the assignment:  
+
+[!code-csharp[DeconstructNames](../../samples/snippets/csharp/new-in-7/new-in-7/program.cs#13_DeconstructNames "Deconstruct with new names")]
+
+Tuples are a lightweight syntax that gives you the ability to declare data-only
+types with multiple fields. You can learn more in depth about them in the
+[tuples topic](tuples.md).
 
 ## Pattern matching
 
+*Pattern Matching* is a feature where you can implement method dispatch on
+properties other than the type of an object. You're probably already familiar
+with method dispatch based on the type of an object. In Object Oriented programming,
+virtual and override methods provide language syntax to implement method dispatching
+based on an object's type. 
+
+Virtual methods and overrides to provide specific implementations of methods for different
```

Comment:
This sentence should be in the previous paragraph, with some revision. "...method dispatching based on an object's type. It allows specific implementations..."